packages/flutter/lib/src/services/image_stream.dart

```diff
@@ -274,3 +288,163 @@ class OneFrameImageStreamCompleter extends ImageStreamCompleter {
+/// 
+/// This class deals with 2 types of frames:
+///  * image frames - image frames of an animated image.
+///  * app frames - frames that the flutter engine is drawing to the screen to
+///    show the app GUI.
+///
+/// For single frame images the stream will only complete once.
+///
+/// For animated images, this class eagerly decodes the next image frame,
+/// and notifies the listeners that a new frame is ready on the first app frame
+/// that is scheduled after the image frame duration has passed.
+///
+/// Scheduling new timers only from scheduled app frames, makes sure we pause
+/// the animation when the app is not visible (as new app frames will not be
+/// scheduled).
+///
+/// See the following timeline example:
+///
+/// | Time | Event                                      | Comment                   |
+/// |------|--------------------------------------------|---------------------------|
+/// | t1   | App frame scheduled (image frame A posted) |                           |
+/// | t2   | App frame scheduled                        |                           |
+/// | t3   | App frame scheduled                        |                           |
+/// | t4   | Image frame B decoded                      |                           |
+/// | t5   | App frame scheduled                        | t5 - t1 < frameB_duration | 
+/// | t6   | App frame scheduled (image frame B posted) | t6 - t1 > frameB_duration |
+///
+class MultiFrameImageStreamCompleter extends ImageStreamCompleter {
+  /// Creates a image stream completer.
+  ///
+  /// Immediately starts decoding the first image frame when the codec is ready.
+  ///
+  /// [codec] is a future for an initialized [ui.Codec] that will be used to
+  /// decode the image.
+  /// [scale] is the linear scale factor for drawing this frames of this image
+  /// at their intended size.
+  MultiFrameImageStreamCompleter({
+    @required Future<ui.Codec> codec,
+    @required double scale,
+    InformationCollector informationCollector
+  }) :
+    _informationCollector = informationCollector,
+    _scale = scale,
+    _framesEmitted = 0,
+    _timer = null
+  {
+    assert(codec != null);
+    codec.then<Null>(_onCodecReady, onError: (dynamic error, StackTrace stack) {
+      FlutterError.reportError(new FlutterErrorDetails(
+        exception: error,
+        stack: stack,
+        library: 'services',
+        context: 'resolving an image codec',
+        informationCollector: informationCollector,
+        silent: true,
+      ));
+    });
+  }
+
+  ui.Codec _codec;
+  final double _scale;
+  final InformationCollector _informationCollector;
+  ui.FrameInfo _nextFrame;
+  // When the current was first shown.
+  Duration _shownTimestamp;
+  // The requested duration for the current frame;
+  Duration _frameDuration;
+  // How many frames have been emitted so far.
+  int _framesEmitted;
+  Timer _timer;
+
+  void _onCodecReady(ui.Codec codec){
```

Comment:
I think _handleCodecReady is probably a more conventional name for this
see https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#naming-rules-for-typedefs-and-function-variables

Reply:
(same with the other _onFoo names below)